src/host.rs

```diff
@@ -51,22 +37,30 @@ impl Host {
             Err(ParseError::EmptyHost)
         } else if input.starts_with("[") {
             if input.ends_with("]") {
-                Ipv6Address::parse(&input[1..input.len() - 1]).map(Host::Ipv6)
+                if let Ok(addr) = input[1..input.len() - 1].parse() {
+                    Ok(Host::V6(addr))
+                } else {
+                    Err(ParseError::InvalidIpv6Address)
+                }
             } else {
                 Err(ParseError::InvalidIpv6Address)
             }
         } else {
-            let decoded = percent_decode(input.as_bytes());
-            let domain = String::from_utf8_lossy(&decoded);
-            // TODO: Remove this check and use IDNA "domain to ASCII"
-            if !domain.is_ascii() {
-                Err(ParseError::NonAsciiDomainsNotSupportedYet)
-            } else if domain.find(&[
-                '\0', '\t', '\n', '\r', ' ', '#', '%', '/', ':', '?', '@', '[', '\\', ']'
-            ][..]).is_some() {
-                Err(ParseError::InvalidDomainCharacter)
+            if let Ok(addr) = input.parse() {
```

Comment:
@pyfisch @Ms2ger have you verified that `Ipv4Addr::parse` and `Ipv6::parse` match https://url.spec.whatwg.org/#host-parsing in every edge case? For example: leading zeros, hex/octal components in v4, …